pvanalytics/features/shading.py

```diff
@@ -0,0 +1,338 @@
+    """Convert a binary image to a boolean series.
+
+    Parameters
+    ----------
+    image : ndarray
+        Binary image.
+    image_index : pd.Index
+        Index for the returned series.
+
+    Returns
+    -------
+    Series
+        A pandas Series with index `index` and values from the image.
+    """
+    return pd.Series(image.flatten(), index=image_index)
+
+
+def _smooth(image):
+    """Smooth `image` by local mean filtering.
+
+    Local means are computed in a 3x3 square surrounding each pixel.
+    """
+    return ndimage.uniform_filter(image, size=15)
+
+
+def _prepare_images(ghi, clearsky, daytime, interval):
+    """Prepare data as images.
+
+    Performs pre-processing steps on `ghi` and `clearsky` before
+    returning images for use in the shadow detection algorithm.
+
+    Returns
+    -------
+    ghi_image : np.ndarray
+        Image form of `ghi`
+    clearsky_image : np.ndarray
+        Image form of `clearsky`
+    clouds_image : np.ndarray
+        Image of the cloudy periods in `ghi`
+    image_times : pandas.DatetimeIndex
+        Index for the data included in the returned images. Leading
+        and trailing days with incomplete data are not included in the
+        image, these times are needed to build a Series from the image
+        later on.
+
+    """
+    # Fill missing days by interpolation. Interpolation is limited to
+    # times with valid data before and after; missing data at the
+    # beginning or end of the series is not filled in.
+    image_width = 1440 // interval
+    ghi = ghi.interpolate(limit_area='inside')
+    # drop incomplete days.
+    ghi = ghi[ghi.resample('D').transform('count') == image_width]
+    image_times = ghi.index
+    ghi_image = _to_image(ghi.to_numpy(), image_width)
+    scaled_ghi = (ghi * 1000) / _smooth(ghi_image).max()
+    scaled_clearsky = (clearsky * 1000) / clearsky.max()
+    scaled_clearsky = scaled_clearsky.reindex_like(scaled_ghi)
+    daytime = daytime.reindex_like(scaled_ghi)
+    # Detect clouds.
+    clouds = _detect_clouds(scaled_ghi, scaled_clearsky, '50T')
+    # Interpolate across days (i.e. along columns) to remove clouds
+    # replace clouds with nans
+    #
+    # This could probably be done directly with scipy.interpolate.inter1d,
+    # but the easiest approach is to turn the image into a dataframe and
+    # interpolate along the columns.
+    cloudless_image = ghi_image.copy()
+    cloudless_image[_to_image(clouds.to_numpy(), image_width)] = np.nan
+    clouds_image = ghi_image.copy()
+    clouds_image[~_to_image(clouds.to_numpy(), image_width)] = np.nan
+    ghi_image = pd.DataFrame(cloudless_image).interpolate(
+        axis=0,
+        limit_direction='both'
+    ).to_numpy()
+    # set night to nan
+    ghi_image[~_to_image(daytime.to_numpy(), image_width)] = np.nan
+    return (
+        ghi_image,
+        _to_image(scaled_clearsky.to_numpy(), image_width),
+        clouds_image,
+        image_times
+    )
```

Comment:
This amount of shuffling between Series and 2D formats makes me think it's an opportunity to improve performance.

Maybe save and re-use `cloud_filter = _to_image(clouds.to_numpy(), image_width)`

Reply:
Might be some room to improve by using scipy interpolation functions as well, avoiding going back to pandas all together. I had a hard time figuring out that API which is why I went back to pandas, do you know how to interpolate along one dimension of an `ndarray`?

Reply:
`ghi_image = pd.DataFrame(` keeps each minute of the day as a column, right?

I'm pretty sure pandas is just wrapping the scipy interpolate functions. Axis-wise interpolation is an [open feature request](https://github.com/scipy/scipy/issues/6187) for scipy.

If the function's performance is not acceptable and this manipulation is the limiting step, we could probably write our own interpolator and skip the pandas and scipy overhead. The result may not be any faster.

OK for now, and we can revisit this point if performance is a problem.